Cython/Compiler/ExprNodes.py

```diff
@@ -1277,6 +1277,9 @@ def coerce_to(self, dst_type, env):
             error(self.pos, "Cannot assign None to %s" % dst_type)
         return super().coerce_to(dst_type, env)
 
+    def analyse_as_type(self, env):
+        return py_none_type
+
```

Comment:
> Doesn't this allow code like `cdef None x` or `x: None = 25` ?

`cdef None` is not allowed even in this PR - I can add test for that if needed:
```
(cython311) root@c113a6388b51:~/dev/cython/test# cat aa.pyx
cdef None x
(cython311) root@c113a6388b51:~/dev/cython/test# cython -3 aa.pyx

Error compiling Cython file:
------------------------------------------------------------
...
cdef None x
     ^
------------------------------------------------------------

aa.pyx:1:5: 'None' is not a type identifier
```
And `x: None` is already allowed in master (I suspect it fallbacks to `object`)

> It seems to me that Optional[] or Union[] might just be special enough to specifically look for None in their type arguments rather than treating None as a type generally. 

We need to have `None` type here to distinguish two cases:
1. `None` value in type definition (e.g. `Union[None, ExtentionType]`) and
2.  unknown type (e.g. `Union[ExtentionType, NonExistingType]`) - in this case we are having `None` value in `specialize_here()` function:

https://github.com/cython/cython/pull/6333/files#diff-7f63ffd84e67d3d113d7ad10fa110111c779a8ca5edd7e12dc51097b36405eabR4773-R4785

> None really isn't a type

I don't fully agree. There is `None` type but is not part of public CAPI [1] of python (and this PR it uses just internal representation of this type - not using python CAPI):
```
>>> type(None)
<class 'NoneType'>
```

In general, I am using None type only to distinguish `None` in type from unknown type. (Moreover it makes sense to me to use type for None) If there is better/other solution I am fine to change the approach.

[1] https://docs.python.org/3/c-api/none.html

Reply:
If we find `None` in code, we don't replace it with `None` as value, but with a `NoneNode`. That makes it easy to distinguish from the value `None` (as in "no node").

As you correctly wrote, the Python type is called `NoneType`, not `None`. The fact that Python uses `None` in the type system instead is simply that for users, typing `NoneType` is very unusual. Many programmers probably have never seen it but know `None` only as a constant value. Thus, `None` is ubiquitous and easy to read and understand as part of a type annotation – it allows `None` as a value, thus making the typed value optional.

However, that doesn't mean that `None` *is* a type, or that we have to represent it as a type in Cython. Just check the type declaration node for `node.is_none` to see if we found `None` in the source code.

Reply:
> However, that doesn't mean that `None` _is_ a type, or that we have to represent it as a type in Cython. Just check the type declaration node for `node.is_none` to see if we found `None` in the source code.

The core of this PR is following part:

```python
    def specialize_here(self, pos, env, template_values=None):
        if py_none_type in template_values:
            self.contains_none = True
```
https://github.com/cython/cython/pull/6333/files#diff-7f63ffd84e67d3d113d7ad10fa110111c779a8ca5edd7e12dc51097b36405eabR4782-R4784

Here, we are detecting whether annotation contains `None` (E.g. `Union[XYZ, None]`). I double checked it and without introducing `NoneType`, this code gets literal `None` value for both cases:

* `Union[int, None]`
* `Union[int, DummyType]`

Hence, without further changes I **cannot** distinguish following cases:
1. non-recognized type - in this case we should evaluate annotation as `object`
2. `None` value - in this case the type is candidate for evaluation of optional type - e.g. `Union[]`

> [!NOTE] 
> To replicate it, just use master and put debug print/breakpoint in following line:
> https://github.com/cython/cython/blob/5805f4c0078e24438edb931a9846f83674c22059/Cython/Compiler/PyrexTypes.py#L4775
> Both cases
> ```python
> import typing
> 
> def main(a: typing.Optional[Bar]):
>     pass
> ```
> and
> ```python
> import typing
> 
> def main(a: typing.Optional[None]):
>     pass
> ```
> contains value `[None]` in variable `template_values`.

Hence, when I want to avoid introducing `NoneType` we need to introduce major structural changes how annotation is evaluated at least in following function (which feeds `specilize_here()`:
https://github.com/cython/cython/blob/5805f4c0078e24438edb931a9846f83674c22059/Cython/Compiler/Nodes.py#L1254

Reply:
> we need to introduce major structural changes how annotation is evaluated at least in following function

Or implement the PR in totally different way, but in that case I have no idea how...

Reply:
> this code gets literal `None` value for both cases:
> 
>     * `Union[int, None]`
> 
>     * `Union[int, DummyType]`

Then we need to make both distinguishable. A way to do that is to introduce `NoneType` as type and replace a literal `None` in type annotations with it.

I'm unsure whether that also applies to the `func() -> None` return type case, where the actual meaning is "has no return value" rather than "returns the constant value None". But it definitely applies to type declarations like the above.

Is there a reason, BTW, why we cannot represent `Union[xyz, None]` as `Optional[xyz]`, which we already support?

Reply:
> Then we need to make both distinguishable. A way to do that is to introduce NoneType as type and replace a literal None in type annotations with it.

Yes, I think this PR is doing pretty that.

> I'm unsure whether that also applies to the func() -> None return type case, where the actual meaning is "has no return value" rather than "returns the constant value None". But it definitely applies to type declarations like the above.

I am not sure if it is important in context of Cython. In this PR Cython generates function which returns `PyObject *` value which is correct (I suppose)

> Is there a reason, BTW, why we cannot represent Union[xyz, None] as Optional[xyz], which we already support?

Honestly, the only reason is to more complete support of PEP 484.